starlite/extensions/rust/route_map.rs

```diff
@@ -0,0 +1,410 @@
+    starlite: &'py PyAny,
+    /// HTTPRoute
+    http_route: &'py PyType,
+    /// WebSocketRoute
+    web_socket_route: &'py PyType,
+    /// ASGIRoute
+    asgi_route: &'py PyType,
+}
+
+/// A node for the trie
+#[derive(Debug, Clone)]
+struct Node {
+    components: HashSet<String>,
+    children: HashMap<String, Node>,
+    path_parameters: Option<Vec<HashMap<String, Py<PyAny>>>>,
+    asgi_handlers: Option<HashMap<String, Py<PyAny>>>,
+    is_asgi: Option<bool>,
+    static_path: Option<String>,
+}
+
+impl Node {
+    /// Creates a new trie node
+    pub fn new() -> Self {
+        Self {
+            components: HashSet::new(),
+            children: HashMap::new(),
+            path_parameters: None,
+            asgi_handlers: None,
+            is_asgi: None,
+            static_path: None,
+        }
+    }
+
+    /// Converts the Rust representation to a PyDict representation
+    pub fn as_pydict(&self) -> PyResult<Py<PyDict>> {
+        let gil = Python::acquire_gil();
+        let dict = PyDict::new(gil.python());
+
+        dict.set_item("components", self.components.clone())?;
+        if let Some(ref asgi_handlers) = self.asgi_handlers {
+            dict.set_item("asgi_handlers", asgi_handlers)?;
+        }
+
+        if let Some(ref path_parameters) = self.path_parameters {
+            dict.set_item("path_parameters", path_parameters)?;
+        }
+        if let Some(is_asgi) = self.is_asgi {
+            dict.set_item("is_asgi", is_asgi)?;
+        }
+        if let Some(ref static_path) = self.static_path {
+            dict.set_item("static_path", static_path)?;
+        }
+
+        Ok(dict.into())
+    }
+}
+
+/// A path router based on a prefix tree / trie.
+///
+/// Stores a handler references and other metadata for each node,
+/// as well as both static paths and plain routes for use with Starlite.
+///
+/// Routes can be added using `add_routes`.
+/// Given a scope containing a path, can retrieve handlers using `parse_scope_to_route`.
+#[pyclass]
+pub struct RouteMap {
+    map: Node,
+    static_paths: HashSet<String>,
+    plain_routes: HashSet<String>,
+}
+
+impl Default for RouteMap {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[pymethods]
+impl RouteMap {
+    /// Creates an empty `RouteMap`
+    #[new]
+    pub fn new() -> Self {
+        RouteMap {
+            map: Node::new(),
+            plain_routes: HashSet::new(),
+            static_paths: HashSet::new(),
+        }
+    }
+
+    /// Adds a new static path by path name
+    pub fn add_static_path(&mut self, path: &str) -> PyResult<()> {
+        self.static_paths.insert(path.to_string());
+        Ok(())
+    }
+
+    /// Checks if a given path refers to a static path
+    pub fn is_static_path(&self, path: &str) -> PyResult<bool> {
+        Ok(self.static_paths.contains(path))
+    }
+
+    /// Removes a path from the static path set
+    pub fn remove_static_path(&mut self, path: &str) -> PyResult<bool> {
```

Comment:
Interesting, but do we need this? Or is this future proofing?

Reply:
Future proofing, wasn't sure if something like this might be necessary, since Python cannot directly access the Rust static path/plain route sets so it doesn't have full control over their data. If you all think this won't be needed I'll get rid of it